src/Mvc/Controller/AbstractRestfulController.php

```diff
@@ -0,0 +1,163 @@
+use ZfrRest\Http\Exception\Client\UnprocessableEntityException;
+
+/**
+ * Base RESTful controller
+ *
+ * Contrary to older versions of ZfrRest, the new controller is much more lightweight, and does a lot less. It
+ * basically dispatches the request to a method based on its HTTP verb. For instance, if you are doing a POST
+ * request, it will dispatch the request to a method called "post" in your controller
+ *
+ * @author  Michaël Gallego <mic.gallego@gmail.com>
+ * @licence MIT
+ */
+abstract class AbstractRestfulController extends AbstractController
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function onDispatch(MvcEvent $event)
+    {
+        $request = $event->getRequest();
+
+        if (!$request instanceof HttpRequest) {
+            throw new RuntimeException('RESTful controller from ZfrRest can only handle HTTP requests');
+        }
+
+        // ZfrRest RESTful controller allows usage of action, in order to avoid controller duplication for things
+        // that do not map well to REST. It does by checking the ":action" parameter. If present, it will use it
+        // like standard action controller
+        if ($action = $this->params('action')) {
+            $method = strtolower($action) . 'Action';
+        } else {
+            $method = strtolower($request->getMethod());
+        }
+
+        if (!method_exists($this, $method)) {
+            throw new MethodNotAllowedException('', null, $this->getAllowedVerbs());
+        }
+
+        $routeParameters = $this->params()->fromRoute(null, []);
+        unset($routeParameters['controller'], $routeParameters['action']);
+
+        $result = $this->$method($routeParameters);
+
+        $event->setResult($result);
+    }
+
+    /**
+     * Provides a built-in implementation for the HTTP OPTIONS method
+     *
+     * It returns the supported methods for the given resource
+     *
+     * @return HttpResponse
+     */
+    public function options()
+    {
+        /** @var HttpResponse $response */
+        $response = $this->getResponse();
+
+        $response->setContent('');
+        $response->setStatusCode(200);
+        $response->getHeaders()->addHeaderLine('Allow', implode(', ', $this->getAllowedVerbs()));
+
+        return $response;
+    }
+
+    /**
+     * @TODO: when we have the new input filter, we should use named validation group and context
+     *
+     * @param  string $inputFilterName
+     * @param  array  $validationGroup
+     * @return array
+     * @throws
+     */
+    protected function validateIncomingData($inputFilterName, array $validationGroup = [])
+    {
+        /** @var \Zend\InputFilter\InputFilterInterface $inputFilter */
+        $inputFilter = $this->serviceLocator->get('InputFilterManager')->get($inputFilterName);
+
+        if (!empty($validationGroup)) {
+            $inputFilter->setValidationGroup($validationGroup);
+        }
+
+        $data = json_decode($this->request->getContent(), true);
```

Comment:
I'm thinking about not allowing other types than JSON. JSON API is the future, supporting XML is not really a good idea. It will complicate implementation for very very few benefit.

Having said this, yes we could maybe have a RequestBodyExtractor or something like this.